feat: Upgrade VisionCamera to V3

mrousavy opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details

Migrates from VisionCamera V2 to VisionCamera V3 ✨ 

Some important changes about V3:

- Fully rewritten codebase for Android (now uses the low-level Camera API "Camera2" in favor of the more limiting alternative "CameraX")
- Fully rewritten Frame Processors implementation (now uses react-native-worklets-core, FPs are disabled if Worklets is not installed; should fix a ton of build issues)
- Fully rewritten Camera Devices API (`useCameraDevice(..)`; `useCameraFormat(..)` - always have a device, never null)
- Muuuuch faster (Camera device selection is now instant, and Camera is faster in general)
- See full changelog here: https://github.com/mrousavy/react-native-vision-camera/compare/v2.16.2...v3.3.0

> [!WARNING]
> Since the Android part is completely rewritten from CameraX to Camera2, there might be some rare edge cases on some phones. Often it's some Samsungs that report a feature as being available in their APIs, then the feature is not available for some reason (e.g. Samsung phones return 60 FPS as a supported FPS range, but just crash when you use 60 FPS. Aka; their APIs simply lie.)
> 
> We should carefully roll this out and report any issues straight to the VisionCamera repo so I can fix them asap! It's a huge help for me to test in a variety of devices.

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$https://github.com/Expensify/App/issues/28341
PROPOSAL: https://github.com/Expensify/App/issues/28341#issuecomment-1748508986


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Open Scan Receipt screen
- [x] Check if Camera works and can take pictures (on iPhones as well as Android phones)
- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

</details>

mrousavy committed:
feat: Upgrade VisionCamera to V3

shubham1206agra commented:
@allroundexperts We need someone to test this with iPhone 15 due to a known issue

shubham1206agra commented:
FYI @allroundexperts @MonilBhavsar 
https://github.com/Expensify/App/pull/27820#issuecomment-1748749252

shubham1206agra commented:
> @shubham1206agra I think that will be tested as part of #27820. No?

To be safe, we should test v3 too.
Once you have tested it, we can ask someone in slack to test this on the iPhone 15

mrousavy commented:
I have an iPhone 15 Pro test device

shubham1206agra commented on src/pages/iou/ReceiptSelector/index.native.js:
```diff
@@ -89,8 +89,7 @@ function getImagePickerOptions(type) {
 function ReceiptSelector({route, report, iou, transactionID, isInTabNavigator}) {
-    const devices = useCameraDevices('wide-angle-camera');
-    const device = devices.back;
+    const device = useCameraDevice('back');
```
Shouldn't we also enable wide-angle-camera too?

mrousavy replied:
Nope, that's already handled by default :)

mrousavy committed:
chore: Bump minSdkVersion to 26

Required for VisionCamera, and will cover 94% of all devices worldwide anyways.

mountiny commented:
Kicked off the test build

OSBotify commented:
Use the links below to test this build in android and iOS. Happy testing!

OSBotify commented:
:test_tube::test_tube: Use the links below to test this build in android and iOS. Happy testing! :test_tube::test_tube:
| android :robot:  | iOS :apple: |
| ------------- | ------------- |
| ❌ FAILED ❌  | ❌ FAILED ❌  |
| The QR code can't be generated, because the android build failed  | The QR code can't be generated, because the iOS build failed  |
| desktop :computer: | web :spider_web: |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/28914/NewExpensify.dmg  | https://28914.pr-testing.expensify.com  |
| ![Desktop](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/28914/NewExpensify.dmg)  | ![Web](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://28914.pr-testing.expensify.com)  |

AndrewGable commented:
Not sure, I just retried it here https://github.com/Expensify/App/actions/runs/6424873394

OSBotify commented:
Use the links below to test this build in android and iOS. Happy testing!

OSBotify commented:
:test_tube::test_tube: Use the links below to test this build in android and iOS. Happy testing! :test_tube::test_tube:
| android :robot:  | iOS :apple: |
| ------------- | ------------- |
| ❌ FAILED ❌  | https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/28914/index.html  |
| The QR code can't be generated, because the android build failed  | ![iOS](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/28914/index.html)  |
| desktop :computer: | web :spider_web: |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/28914/NewExpensify.dmg  | https://28914.pr-testing.expensify.com  |
| ![Desktop](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/28914/NewExpensify.dmg)  | ![Web](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://28914.pr-testing.expensify.com)  |

AndrewGable commented:
Seems like Android is failing to build on the GitHub runner, are you seeing the same locally?

https://github.com/Expensify/App/actions/runs/6424873394/job/17446462378#step:10:10762

![Screenshot 2023-10-05 at 4 30 35 PM](https://github.com/Expensify/App/assets/2838819/f1e291f7-49ec-44bd-bc44-ece9a516363d)

shubham1206agra commented:
@mrousavy Can you check this line?
https://github.com/mrousavy/react-native-vision-camera/blob/36e9e00930cac75c5890a2f8dc688874425def57/package/android/gradle.properties#L17

mrousavy commented:
Hi - this is the error:

![image](https://github.com/Expensify/App/assets/15199031/0d59db81-52a5-47fa-b4ee-2ac6bd6bba0a)

From the docs it looks like this has been available since Kotlin 1.5, so not sure if 1.7 is really required. I'll try with 1.7

mrousavy commented:
Either we need to add `apply plugin: 'kotlin-android'` to the `build.gradle`, or we need to upgrade Kotlin.

I believe the funcs I'm using are in the std lib of Kotlin, this is no third party extensions dep or anything afaik.

mrousavy committed:
fix: Update kotlinVersion

mrousavy committed:
Update Podfile.lock

mrousavy committed:
chore: Remove AsyncStorage (it was unused)

mrousavy commented:
Update:

1. I just updated the kotlinVersion from 1.6.x to 1.7.x because we might need that for VisionCamera
2. Since react-native-worklets-core is not installed, Frame Processors are automatically disabled now:
    ![Screenshot 2023-10-09 at 11 30 08](https://github.com/Expensify/App/assets/15199031/e3028442-6e5c-4f76-8e8a-50a49a029deb)
3. I have removed @react-native-async-storage from the dependencies because it is no longer used and the build was failing because of it. We now use Quick-SQLite/IndexDB under the hood for Onyx.

Since Cameras are hardware features and VisionCamera V3 now uses the low-level Camera2 API, we def. need to be quick with feedback here - I just know there's gonna be some weird Samsungs or something that break for some reason.

If there are any issues with VisionCamera V3, just ping me on Slack or create an issue with details on the VisionCamera issues page!

mrousavy commented:
I got some issues on my end, will try to fix them asap to get the build up and running
![Screenshot 2023-10-09 at 14 08 33](https://github.com/Expensify/App/assets/15199031/1957908b-0f17-49a7-af34-9dcfbc5d957c)

OSBotify commented:
:test_tube::test_tube: Use the links below to test this build in android and iOS. Happy testing! :test_tube::test_tube:
| android :robot:  | iOS :apple: |
| ------------- | ------------- |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/android/28914/index.html  | https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/28914/index.html  |
| ![Android](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/android/28914/index.html)  | ![iOS](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/28914/index.html)  |
| desktop :computer: | web :spider_web: |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/28914/NewExpensify.dmg  | https://28914.pr-testing.expensify.com  |
| ![Desktop](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/28914/NewExpensify.dmg)  | ![Web](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://28914.pr-testing.expensify.com)  |

mrousavy commented:
@allroundexperts thanks for testing, I need the native logs to see where this is coming from though.
Can you please share the adb logcat logs here?

mrousavy commented:
@allroundexperts I can see that, but I still need the logs. I don't have a Galaxy A13 phone myself.

If you don't know how to get native logs:
1. Connect your phone via USB
2. Run `adb logcat`
3. Start doing what you just did to reproduce the crash
4. Copy the logs from the beginning to the end
5. Paste them here (under a collapsable details tag or via a gist)

If your phone doesn't allow logcat streaming in release it would be great if you could just quickly build the app from this branch in debug and try getting the logs that way.

mrousavy commented:
> Can't you reproduce this one on your side?

No, on my Huawei P10 it does not crash, even if I navigate back and forth multiple times.

mountiny reviewed (dismissed):
@allroundexperts @MonilBhavsar @AndrewGable I think we should go ahead with the PR so we get more testing in on staging as we dont have that much time before the conferences where we need the camera to work

mountiny commented:
I have tested the build on iphone 14 pro and works like a charm

mountiny commented:
I think it might be easier to get this to staging and let QA test with their phones and then we will get better picture of whats happening and address it then, but obviously if you want to go ahead with that plan that works too

mrousavy commented:
Okay, but I need the adb logs, does your QA team have those?

mountiny commented:
> Okay, but I need the adb logs, does your QA team have those?

I dont think so, we get firebase crashlytics logs at least. @allroundexperts is travelling today so not sure if he can get those logs. I dont have any android device either

mrousavy commented:
This might fix the issue, will test more: https://github.com/mrousavy/react-native-vision-camera/pull/1972

puneetlath commented:
@mountiny do you think we should go ahead and merge this and get it to staging? Ahead of money 2020?

mrousavy commented:
Hey - I'm still working on making sure the Android crash never occurs again, I work on this in the VisionCamera repo and will merge to main & release once this PR is ready: https://github.com/mrousavy/react-native-vision-camera/pull/1975, then I'll update the Expensify PR and we should be good to merge :)

mountiny commented:
@puneetlath @AndrewGable I believe given the tight turnaround we should rather merge this v2 bump https://github.com/Expensify/App/pull/27820/commits/033d7fd93e41049c11c3d417714fa2b8df5cb02e as we need this out working fro the latest iphones

MonilBhavsar commented:
Agree to the plan 👍

mrousavy committed:
Update VisionCamera to 3.4.0 (which includes a fix we need)

mrousavy committed:
Update lockfiles

mrousavy commented:
I updated this PR now - rebased with main, and upgraded to VisionCamera 3.4.0.
This should include some stability fixes on Android, and huge performance improvements on iOS.

I'll test to see if the Android bug has been fixed now.
If not, we need this VisionCamera PR: https://github.com/mrousavy/react-native-vision-camera/pull/1996

mountiny commented:
Triggered a fresh build https://github.com/Expensify/App/actions/runs/6549536809

OSBotify commented:
:test_tube::test_tube: Use the links below to test this build in android and iOS. Happy testing! :test_tube::test_tube:
| android :robot:  | iOS :apple: |
| ------------- | ------------- |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/android/28914/index.html  | ❌ FAILED ❌  |
| ![Android](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/android/28914/index.html)  | The QR code can't be generated, because the iOS build failed  |
| desktop :computer: | web :spider_web: |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/28914/NewExpensify.dmg  | https://28914.pr-testing.expensify.com  |
| ![Desktop](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/28914/NewExpensify.dmg)  | ![Web](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://28914.pr-testing.expensify.com)  |

mrousavy commented:
Also, iOS build fails because the CI is not on Xcode 15 yet - should I add special handling if the user is not yet on Xcode 15 or will the CI be upgraded soon? 

I'm using new APIs from iOS 17, feels like we need to upgrade that in the coming days anyways...

AndrewGable commented:
We can upgrade to XCode 15/iOS 17, but not sure it should be done in this PR.

mountiny commented:
I agree we should do this in different PR

mrousavy commented:
Cool yup no problem, Xcode 15 upgrade can wait I added extra ifdefs to VisionCamera so it isnt required :)

mrousavy commented:
Yes I know, I didn't fix that yet. I'm working on it later today or tomorrow

mrousavy committed:
chore: Bump VisionCamera to 3.5.0

mrousavy committed:
fix: Patch VisionCamera to fix the blackscreen when navigating back

I'm working on a real fix in VisionCamera in the meantime :)

mrousavy commented:
@allroundexperts @mountiny @AndrewGable just created a hot-fix for the blackscreen issue. I identified the problem and can confirm that it will be fixed with this PR: https://github.com/mrousavy/react-native-vision-camera/pull/1996

I pushed that VisionCamera patch here, so it should work fine now (confirmed on my Huawei something), but I'll now switch to VisionCamera to build a true solid fix for this. :) 

Do we want to wait for that VisionCamera update, or do we want to merge with that hot-patch?

mrousavy commented:
I think it'd be better to wait for the true VisionCamera fix. Just give me a few days and that's done :)

mrousavy commented:
I think I got https://github.com/mrousavy/react-native-vision-camera/pull/2049 working pretty good - that fixes a ton of concurrency issues since now finally everything is happening under an atomic mutex 😄 

on Android this was especially tricky because of the PreviewView since that could just go out of sync from time to time. I'll test more to see if this is still an issue or if everything is fixed now :)


If everything's fixed, I can release this and we should be good to go with the upgrade here in this PR

puneetlath commented:
nice!

mrousavy committed:
fix: Upgrade VisionCamera to 3.6.1 (includes Android blackscreen fix)

mrousavy committed:
fix: Update to 3.6.2. which includes another fix

mrousavy committed:
Merge branch 'main' into upgrade-visioncamera-v3

mrousavy committed:
Delete react-native-vision-camera+3.5.0.patch

mrousavy commented:
Okay, I pushed the new changes and now we are using VisionCamera 3.6.x, which contains the new atomically single-lock `core/` API on Android.

Can we do another test build to gather some feedback? This is pretty early since I just released the new library, but from my testing so far it looks quite good.

cc @puneetlath @mountiny @AndrewGable

mountiny commented:
build running here https://github.com/Expensify/App/actions/runs/6628563393

OSBotify commented:
:test_tube::test_tube: Use the links below to test this build in android and iOS. Happy testing! :test_tube::test_tube:
| android :robot:  | iOS :apple: |
| ------------- | ------------- |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/android/28914/index.html  | https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/28914/index.html  |
| ![Android](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/android/28914/index.html)  | ![iOS](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/28914/index.html)  |
| desktop :computer: | web :spider_web: |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/28914/NewExpensify.dmg  | https://28914.pr-testing.expensify.com  |
| ![Desktop](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/28914/NewExpensify.dmg)  | ![Web](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://28914.pr-testing.expensify.com)  |

Julesssss commented on android/build.gradle:
```diff
@@ -3,7 +3,7 @@
     ext {
         buildToolsVersion = "33.0.0"
-        minSdkVersion = 21
+        minSdkVersion = 26
```
Woah, this is a problem. We'd be blocking 6.8% of users from using the app... That doesn't seem worth it for an improved camera feature. 

What specifically requires that we use a platform min of Android 8?

<img width="556" alt="254630961-b427f24b-c3cb-4a99-b160-e43021cfd0f5" src="https://github.com/Expensify/App/assets/10736861/c522158c-0222-4397-9db2-592a889fab0c">

Julesssss replied:
I see the following listed as high-level new features with Android SDK 26: 

<img width="526" alt="Screenshot 2023-10-25 at 09 19 12" src="https://github.com/Expensify/App/assets/10736861/c01dc3c8-a757-4edf-b2f3-6600401f1304">

Julesssss reviewed (changes requested):
We need to discuss [this](https://github.com/Expensify/App/pull/28914/files#r1371340736). Merging the PR as is will block just under 7% of potential users from using the app.

puneetlath commented:
Created a new issue for this here: https://github.com/Expensify/App/issues/30732

mrousavy commented:
update: I just got a prototype running on my end that works on API 21, so we don’t need to drop the minSdk requirement! :muscle:

puneetlath commented:
Nice!

mrousavy commented:
Important update, I created a new PR: https://github.com/Expensify/App/pull/37483

mountiny commented:
Can we close this in favour of https://github.com/Expensify/App/pull/37483

mrousavy commented:
Ja 👍
